rfd/0094-kubernetes-node-joining.md

```diff
@@ -0,0 +1,335 @@
+- teleport does not support joining from certs (unless we manually build a
+  valid state)
+- tBot has to dynamically discover how many pods want certs and which certs
+  they need (we don't have a streaming destination API yet)
+- pods should have a way to access their specific state:
+    - we can't mount a different secret in each pod
+    - it can be done through admission webhooks
+    - pods can read secrets (this works OK for Teleport nodes, but cannot be
+      generalized with other workloads like access plugin because this requires
+      being able to read all certs from the namespace)
+    - pods can request a cert to tBot over HTTP
+- tBot has to validate the identity of the pod requesting the certificate
+- tBot has to join the cluster initially, which is almost the same problem that
+  we're trying to address (almost because we have less constraints due to
+  autoscaling). The easiest approach would be to bundle tBot with auth nodes and
+  use the local admin socket to bootstrap it, like what we're doing with the
+  operator
+
+#### Approach 3 - Operator/MachineID-generated short-lived tokens
+
+In this approach, short-lived static tokens are generated by the Teleport
+operator and injected in containers using mutation webhooks.
+
+```mermaid
+sequenceDiagram
+    StatefulSet->>Kubernetes: New pod
+    Kubernetes->>+Operator: Mutation webhook
+    Operator->>Teleport Cluster:New token, ttl = 2min
+    Teleport Cluster->>Operator: Token
+    Operator->>Kubernetes: Create secret with token
+    Operator->>-Kubernetes: Mutated pod with secret mount
+    Kubernetes->>Proxy: Start
+    Note over Operator: Wait for token expiry
+    Operator->>Kubernetes: Delete secret
+```
+
+The operator internally relies on MachineID to get its identity, this solution
+is basically an automated version of [MachineID token
+generation](https://github.com/gravitational/teleport/discussions/15404).
+
+This approach brings the following challenges:
+- the operator becomes mandatory, which might not be a possibility for certain
+  users (because of new additional RBAC)
+- the operator enters the critical path, if it is not available no pod can start
+- pods started before the mutation webhook is added will not get their token
+- there is a race condition if the proxy pod does not start before expiry, the
+  failure mode is a pod stuck in crashloop and the only way out is to delete
+  the pod. This is particularly likely to happen on clusters with node
+  autoscaling or capacity issues.
+
+### Choosing the approach
+
+The second approach tends to bring more issues than it solves. Even if we added
+MachineID support for Teleport nodes, this would not be used by customers with
+their Teleport workload.
+
+The third approach adds new components in the critical path and new complex
+failure modes.
+
+The first approach seems the more robust, leverages previous work with
+joinTokens and provides very solid security guarantees for Kubernetes 1.20+.
```

Comment:
```suggestion
ProvisionTokens and provides very solid security guarantees for Kubernetes 1.20+.
```

I think using `ProvisionToken` is more specific terminology (this is the resource name in Teleport) here and avoids conflation with other "tokens" - we have so many tokens 😱